Force line-buffering for stderr

Trott opened the pull request:
Fixes: https://github.com/nodejs/node/issues/2476
Refs: https://github.com/nodejs/node/pull/3615

On CI, it appears that stderr on SmartOS is not line buffered and that was resulting in some interleaving in test-debug-signal-cluster, causing the test to be flaky. This change forces stderr to be line-buffered, fixing the issue.

Trott commented:
(Is there a better tag/prefix/whatever for this than `src:`? `node:`?)

bnoordhuis commented on src/node.cc:
```diff
@@ -3129,6 +3129,7 @@ void LoadEnvironment(Environment* env) {
   // thrown during process startup.
   try_catch.SetVerbose(true);
 
+  setvbuf(stderr , NULL , _IOLBF , 1024);
```
This is not the right place for a call to setvbuf, it's too late after startup.  It should probably be the first thing in main() or `node::Init()` and even that is possibly not soon enough because of printf calls from static constructors.

Trott replied:
On a separate testing branch, I moved it to `main()`.

I had it run the test on SmartOS 500 times and there were no failures. https://ci.nodejs.org/job/node-test-commit-smartos/246/

Then I commented out the line.
https://github.com/nodejs/node/commit/bf88e85653bc27ee8ef8c5a26b54dcff7d641851

And ran the test again and got multiple failures.
https://ci.nodejs.org/job/node-test-commit-smartos/247/

I'll move the `setvbuf()` to `main()` or anywhere else that seems more appropriate.

Trott commented:
CI: https://ci.nodejs.org/job/node-test-pull-request/679/

Trott commented:
CI: https://ci.nodejs.org/job/node-test-pull-request/683/

Trott commented:
CI: https://ci.nodejs.org/job/node-test-pull-request/684/

mscdex commented on src/node_main.cc:
```diff
@@ -41,6 +41,7 @@ int wmain(int argc, wchar_t *wargv[]) {
 #else
 // UNIX
 int main(int argc, char *argv[]) {
+  setvbuf(stderr , NULL , _IOLBF , 1024);
```
I think there shouldn't be a space immediately after each of the first three arguments?

Trott replied:
Clearly, I need to up my cut-and-paste game. Extraneous spaces removed. Thanks!

JungMinu commented:
LGTM

jbergstroem commented:
@trott should probably be `node: $foo`.

Trott commented:
@jbergstroem Cool. Changed the commit message, squashed the commits into a single commit and force pushed. 

Anyone feel comfortable giving this one a thumbs up (assuming CI comes back green)?

jasnell commented:
hmm... would this need to be semver-minor or higher?

Trott commented:
@jasnell I don't think so. I think it only affects messages sent by the Node debugger which bypasses `console.error()` and uses the raw `stderr` instead.  While in theory, `console.error()` should be affected by this, practical tests seem to indicate that it behaves line-buffered regardless. Perhaps due to something on the streams level? I didn't look too deeply at it.

Anyway, if it increased your comfort level to have it at `semver-minor`, I have no objection to that. But if you're asking my opinion, I think it's patch level. I just don't hold that opinion particularly strongly.

bnoordhuis commented:
LGTM.  This is semver-patch IMO, it only affects fprintf statements in src/, not `process.stderr.write` statements in JS land.

jasnell commented:
:+1: LGTM

Trott commented:
Kewl. I'm waiting on the return of the CI server to do one last rebased-against-master CI run before landing...

jbergstroem commented:
LGTM

jasnell commented:
New CI Run: https://ci.nodejs.org/job/node-test-pull-request/689/

Trott commented:
OK to land despite OSX + Arm builbot problems on CI?

jbergstroem commented:
@trott if you've tested locally on a mac I'd say go for it. I'm +1 to patchlevel.

Trott commented:
Landed in 0966ab99966b7d3fbe4d7b93797fb299595fca72.

TheAlphaNerd commented:
should this land in LTS?

/cc @jasnell @bnoordhuis @Trott

rvagg commented:
hey, this is a good one to discuss at an LTS WG meeting—the change itself is a good one but who knows what weird use is out there that may be depending on existing behaviour, perhaps it's not worth the trouble to backport?

TheAlphaNerd commented:
looks like we missed talking about this at todays meeting. @jasnell should we add a label for LTS discussions?

jasnell commented:
Yeah, if you'd like go ahead and create an lts-agenda label for this repo.
On Dec 14, 2015 4:22 PM, "Myles Borins" <notifications@github.com> wrote:

> looks like we missed talking about this at todays meeting. @jasnell
> <https://github.com/jasnell> should we add a label for LTS discussions?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/nodejs/node/pull/3701#issuecomment-164602328>.
>

jasnell commented:
This is flagged as lts-watch but needs to be discussed before landing...

TheAlphaNerd commented:
@nodejs/lts can you please weigh in on this one?

rvagg commented:
From these two comments 

> @jasnell I don't think so. I think it only affects messages sent by the Node debugger which bypasses console.error() and uses the raw stderr instead. 

> LGTM. This is semver-patch IMO, it only affects fprintf statements in src/, not process.stderr.write statements in JS land.

If you `grep fprintf src/*` the impact is pretty low. So +1 from me for LTS.

jasnell commented:
LGTM to v4.4
